includes/services/extended/nextdoor.php

```diff
@@ -0,0 +1,176 @@
+
+
+	function __construct() {
+		parent::__construct();
+
+		// Enable "basic" UI for entering key/secret
+		if ( ! KEYRING__HEADLESS_MODE ) {
+			add_action( 'keyring_nextdoor_manage_ui', array( $this, 'basic_ui' ) );
+			add_filter( 'keyring_nextdoor_basic_ui_intro', array( $this, 'basic_ui_intro' ) );
+		}
+
+		$this->set_endpoint("access_token", "https://auth.nextdoor.com/v2/token", "POST" );
+        $this->set_endpoint("authorize", "https://nextdoor.com/v3/authorize/?scope=openid%20post:write&", "POST");
+
+		$creds = $this->get_credentials();
+		if ( is_array( $creds ) ) {
+			$this->app_id = $creds['app_id'];
+			$this->key    = $creds['key'];
+			$this->secret = $creds['secret'];
+		}
+		$admin_url          = Keyring_Util::admin_url();
+		$this->redirect_uri = substr( $admin_url, 0, strpos( $admin_url, '?' ) );
+
+		// Send authorization via header
+		$this->authorization_header = 'Bearer';
+		add_filter( 'keyring_nextdoor_request_token_params', array( $this, 'request_token_params' ) );
+		add_filter( 'keyring_nextdoor_verify_token_params', array( $this, 'verify_token_params' ) );
+		add_filter( 'keyring_nextdoor_verify_token_post_params', array( $this, 'verify_token_post_params' ) );
+		add_filter( 'keyring_access_token', array( $this, 'fix_access_token_meta' ), 10, 2 );
+	}
+
+	function basic_ui_intro() {
+		/* translators: url */
+		echo '<p>' . sprintf( __( 'To use the Nextdoor service you need to be manually approved. Please reach out to get your client id and client secret.', 'keyring' )) . '</p>';
+	}
+
+	function request_token_params( $params ) {
+		$url_components = parse_url( $params['redirect_uri'] );
+		parse_str( $url_components['query'], $redirect_state );
+		$redirect_state['state'] = $params['state'];
+		$params['state']         = Keyring_Util::get_hashed_parameters( $redirect_state );
+		$params['redirect_uri']  = $this->redirect_uri;
+
+		return $params;
+	}
+
+	function verify_token_params( $params ) {
+		unset( $params['client_id'] );
+		unset( $params['client_secret'] );
+		$params['redirect_uri']  = $this->redirect_uri;
+		// $params[''] = ($params['code'] . '=');
+		return $params;
+	}
+
+	function verify_token_post_params( $params ) {
+		$params['headers'] = $this->get_basic_auth();
+
+		return $params;
+	}
+
+	function fix_access_token_meta( $access_token, $token ) {
+		error_log('testing fix access token meta');
+		if ( 'nextdoor' !== $access_token->get_name() ) {
+			return $access_token;
+		}
+
+		return new Keyring_Access_Token(
+			$this->get_name(),
+			$token['access_token'],
+			array_merge( $access_token->get_meta(), $this->get_token()->get_meta() ) // refresh_token has been updated, and we want to make sure we store it
+		);
+	}
+
+	function get_basic_auth() {
+		return array( 'Authorization' => 'Basic ' . base64_encode( $this->key . ':' . $this->secret ) );
+	}
+
+	function build_token_meta( $token ) {
+		$decoded_id_token = json_decode(base64_decode(explode('.', $token['id_token'])[1]));
```

Comment:
Yeah, I'm using `explode` to parse the `id_token` since we only wanted to grab the payload and not the `header` and `signature` of the token. 

https://stackoverflow.com/questions/31146111/how-do-i-extract-the-payload-data-of-a-decoded-jwt-web-token-using-php

this looks promising

Reply:
Sure! It might just be a good idea to check that we have `id_token` and at least 2 elements after the `explode` etc. Although we might be able to guarantee all that before we get here.